### DISCUSSION TITLE
ceph-volume lvm batch is not creating OSDs on partitions in latest Nautilus v14.2.13 and Octopus v15.2.8

### DISCUSSION THREAD

**Discussion by travisn** [rook/rook#7349] [General]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
In the latest Nautilus v14.2.15 and Octopus 15.2.8, `ceph-volume lvm batch` is not allowing an OSD to be created on a raw partition.

In the integration tests we are seeing this with the following in the osd prepare job:
```
2020-12-17 06:43:11.875525 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --osds-per-device 1 /dev/sda1 --report
2020-12-17 06:43:12.515146 D | exec: usage: ceph-volume lvm batch [-h] [--db-devices [DB_DEVICES [DB_DEVICES ...]]]
2020-12-17 06:43:12.515335 D | exec:                              [--wal-devices [WAL_DEVICES [WAL_DEVICES ...]]]
2020-12-17 06:43:12.515413 D | exec:                              [--journal-devices [JOURNAL_DEVICES [JOURNAL_DEVICES ...]]]
2020-12-17 06:43:12.515480 D | exec:                              [--auto] [--no-auto] [--bluestore] [--filestore]
2020-12-17 06:43:12.515619 D | exec:                              [--report] [--yes]
2020-12-17 06:43:12.515976 D | exec:                              [--format {json,json-pretty,pretty}] [--dmcrypt]
2020-12-17 06:43:12.516441 D | exec:                              [--crush-device-class CRUSH_DEVICE_CLASS]
2020-12-17 06:43:12.521079 D | exec:                              [--no-systemd]
2020-12-17 06:43:12.521237 D | exec:                              [--osds-per-device OSDS_PER_DEVICE]
2020-12-17 06:43:12.521308 D | exec:                              [--data-slots DATA_SLOTS]
2020-12-17 06:43:12.521394 D | exec:                              [--block-db-size BLOCK_DB_SIZE]
2020-12-17 06:43:12.521460 D | exec:                              [--block-db-slots BLOCK_DB_SLOTS]
2020-12-17 06:43:12.521524 D | exec:                              [--block-wal-size BLOCK_WAL_SIZE]
2020-12-17 06:43:12.521575 D | exec:                              [--block-wal-slots BLOCK_WAL_SLOTS]
2020-12-17 06:43:12.521656 D | exec:                              [--journal-size JOURNAL_SIZE]
2020-12-17 06:43:12.521738 D | exec:                              [--journal-slots JOURNAL_SLOTS] [--prepare]
2020-12-17 06:43:12.521813 D | exec:                              [--osd-ids [OSD_IDS [OSD_IDS ...]]]
2020-12-17 06:43:12.521880 D | exec:                              [DEVICES [DEVICES ...]]
2020-12-17 06:43:12.521930 D | exec: ceph-volume lvm batch: error: /dev/sda1 is a partition, please pass LVs or raw block devices
```

**Expected behavior:**
Raw partitions have been working and expected to continue working.

**How to reproduce it (minimal and precise):**

Attempt to create an OSD on a partition with v15.2.8.

**Comment by travisn** [comment]
Looks related to this change in c-v: https://github.com/ceph/ceph/pull/38280

**Comment by pavanfhw** [comment]
Runned into this issue. What should I do for now?
Do I need to change the Ceph version in the cluster deployment? What is the latest workign image?

**Comment by travisn** [comment]
If you need to create OSDs on partitions, you'll need to use Ceph v14.2.12 or v15.2.7 while we are following up on the issue.

**Reply by lbogdan** [reply]
I think this should probably be marked as answer?

**Reply by leseb** [reply]
Because we converted this from an issue, we don't do see "answer" button, at least I don't...

**Comment by leseb** [comment]
Not sure what is the right fix for this in 1.5, https://github.com/rook/rook/pull/4879 is appealing but has its own limitations...

**Comment by haslersn** [comment]
I'm also affected by this problem. Is there a workaround?

**Comment by valkmit** [comment]
I am also looking for a potential fix for this!

**Comment by leseb** [comment]
@hasheddan @valkmit what about this? https://github.com/rook/rook/issues/6849#issuecomment-748314135 is it not feasible for you?

**Comment by haslersn** [comment]
@leseb with v15.2.7 I get a different error. [Full log](https://gist.github.com/haslersn/5b83690dcee7db2f5b44c329b9dbeb4f). Tail of the log:

```
2021-02-15 19:07:33.856933 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --dmcrypt --osds-per-device 1 /dev/sdd /dev/sda /dev/sdb /dev/sdc --db-devices /dev/disk/by-path/pci-0000:11:00.0-nvme-1-part3 --report
2021-02-15 19:07:35.856836 D | exec: Traceback (most recent call last):
2021-02-15 19:07:35.856884 D | exec:   File "/usr/sbin/ceph-volume", line 11, in <module>
2021-02-15 19:07:35.856889 D | exec:     load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
2021-02-15 19:07:35.856894 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 40, in __init__
2021-02-15 19:07:35.856898 D | exec:     self.main(self.argv)
2021-02-15 19:07:35.856903 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
2021-02-15 19:07:35.856907 D | exec:     return f(*a, **kw)
2021-02-15 19:07:35.856911 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 151, in main
2021-02-15 19:07:35.856915 D | exec:     terminal.dispatch(self.mapper, subcommand_args)
2021-02-15 19:07:35.856919 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
2021-02-15 19:07:35.856923 D | exec:     instance.main()
2021-02-15 19:07:35.856927 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/main.py", line 42, in main
2021-02-15 19:07:35.856931 D | exec:     terminal.dispatch(self.mapper, self.argv)
2021-02-15 19:07:35.856934 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
2021-02-15 19:07:35.856938 D | exec:     instance.main()
2021-02-15 19:07:35.856941 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
2021-02-15 19:07:35.856945 D | exec:     return func(*a, **kw)
2021-02-15 19:07:35.856949 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 322, in main
2021-02-15 19:07:35.856952 D | exec:     self._get_explicit_strategy()
2021-02-15 19:07:35.856956 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 332, in _get_explicit_strategy
2021-02-15 19:07:35.856960 D | exec:     self._filter_devices()
2021-02-15 19:07:35.856963 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 385, in _filter_devices
2021-02-15 19:07:35.856967 D | exec:     raise RuntimeError(err.format(len(devs) - len(usable)))
2021-02-15 19:07:35.856973 D | exec: RuntimeError: 1 devices were filtered in non-interactive mode, bailing out
failed to configure devices: failed to initialize devices: failed ceph-volume report: exit status 1
```

**Comment by leseb** [comment]
> @leseb with v15.2.7 I get a different error. [Full log](https://gist.github.com/haslersn/5b83690dcee7db2f5b44c329b9dbeb4f). Tail of the log:
> 
> ```
> 2021-02-15 19:07:33.856933 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --dmcrypt --osds-per-device 1 /dev/sdd /dev/sda /dev/sdb /dev/sdc --db-devices /dev/disk/by-path/pci-0000:11:00.0-nvme-1-part3 --report
> 2021-02-15 19:07:35.856836 D | exec: Traceback (most recent call last):
> 2021-02-15 19:07:35.856884 D | exec:   File "/usr/sbin/ceph-volume", line 11, in <module>
> 2021-02-15 19:07:35.856889 D | exec:     load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
> 2021-02-15 19:07:35.856894 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 40, in __init__
> 2021-02-15 19:07:35.856898 D | exec:     self.main(self.argv)
> 2021-02-15 19:07:35.856903 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
> 2021-02-15 19:07:35.856907 D | exec:     return f(*a, **kw)
> 2021-02-15 19:07:35.856911 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 151, in main
> 2021-02-15 19:07:35.856915 D | exec:     terminal.dispatch(self.mapper, subcommand_args)
> 2021-02-15 19:07:35.856919 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
> 2021-02-15 19:07:35.856923 D | exec:     instance.main()
> 2021-02-15 19:07:35.856927 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/main.py", line 42, in main
> 2021-02-15 19:07:35.856931 D | exec:     terminal.dispatch(self.mapper, self.argv)
> 2021-02-15 19:07:35.856934 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
> 2021-02-15 19:07:35.856938 D | exec:     instance.main()
> 2021-02-15 19:07:35.856941 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
> 2021-02-15 19:07:35.856945 D | exec:     return func(*a, **kw)
> 2021-02-15 19:07:35.856949 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 322, in main
> 2021-02-15 19:07:35.856952 D | exec:     self._get_explicit_strategy()
> 2021-02-15 19:07:35.856956 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 332, in _get_explicit_strategy
> 2021-02-15 19:07:35.856960 D | exec:     self._filter_devices()
> 2021-02-15 19:07:35.856963 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 385, in _filter_devices
> 2021-02-15 19:07:35.856967 D | exec:     raise RuntimeError(err.format(len(devs) - len(usable)))
> 2021-02-15 19:07:35.856973 D | exec: RuntimeError: 1 devices were filtered in non-interactive mode, bailing out
> failed to configure devices: failed to initialize devices: failed ceph-volume report: exit status 1
> ```

I don't know what's going on, could please open a bug here? https://tracker.ceph.com/ under the ceph-volume component?
Thanks

**Comment by haslersn** [comment]
@leseb Thanks, I will do so after my registration is approved.

Could it be that it's not allowed to have metadata pools on partitions? See https://tracker.ceph.com/issues/47966#note-3

**Comment by leseb** [comment]
> @leseb Thanks, I will do so after my registration is approved.
> 
> Could it be that it's not allowed to have metadata pools on partitions? See https://tracker.ceph.com/issues/47966#note-3

Could be but I'm not sure, c-v's partition support is still confusing.

**Comment by DjangoCalendar** [comment]
@travisn does this issue affects upgrades as well.

Let's say I am using:
- ceph:v15.2.7
- rook/ceph:v1.5.3

with this setup I am running on raw partitions.

Will I be able to upgrade to version for example:
- ceph:v15.2.8
- rook/ceph:v1.5.7

Asking in shorter way:
Does this is issue affects only fresh deployment or upgrades as well ?

**Comment by cbartz** [comment]
@DjangoCalendar You can upgrade and use your existing OSD's. But if you want to setup a fresh OSD it wont work on partitions.

**Comment by mathieu-rossignol** [comment]
Hi,

I have rook-ceph-v1.5.8 and I'm getting this issue if I use anything else than ceph/ceph:v15.2.7 image in my CephCluster definition.
At least with these versions of ceph I'm facing this issue:

- v15
- v15.2.8
- v15.2.9

I seems there is a regression after 15.2.7 :(. But in fact isn't it a pure ceph issue ? To summarize for others coming here: **use ceph/ceph:v15.2.7 for the CephCluster image**:

![image](https://user-images.githubusercontent.com/18575218/109657181-7e8cc480-7b65-11eb-9a50-181eee1a5f47.png)

**Comment by leseb** [comment]
Let me summarize (hopefully) once and for all.

**Problem**:  ceph-volume lvm is not creating OSDs on partitions anymore **as of** Nautilus v14.2.13 and Octopus v15.2.8 and **onward**. The error can be seen from the osd prepare job logs and resemble:

```
2021-02-25 09:01:42.471002 D | exec: ceph-volume lvm batch: error: /dev/sda1 is a partition, please pass LVs or raw block devices
```

**Solutions**: 

1. Use full devices (e,g: `/dev/sda`) or Logical Volumes on top of your partition, Rook could happily propose LVs to ceph-volume but passing `dm-` does not seem to work on 15.2.9, see: https://tracker.ceph.com/issues/49582
2. For the time being, **stick** with v14.2.12 or v15.2.7, and once Rook 1.6 is released, upgrade to **at least** v14.2.14 or v15.2.9 to get partitions working again. As of 1.6, Rook OSD's implementation for simple scenarios (one OSD = one disk basically) is not using LVM but RAW mode from ceph-volume. See https://github.com/rook/rook/pull/4879 for more details.

I'm converting to a Discussion so people can continue discussing but highlight the latest solution better.

**Reply by vasanthaganeshk** [reply]
I'm facing the same issue in ceph-version: 15.2.9, rook version: 1.5.9

**Comment by travisn** [comment]
@leseb @BlaineEXE Since we reverted to use lvm mode again instead of raw mode due to the atari partition issue, it seems partitions cannot be used again.

**Reply by BlaineEXE** [reply]
How are you determining that? Partitions are not affected by the Atari bug, and so I never removed the ability to create raw osds on partitions. What may be causing this is an environment that is only partly partitions. If there are any full disks used, Rook will use lvm mode. We need to rework the lvm/raw support to be able to detect which to use on a per-disk basis rather than lvm/raw on a whole node at a time.

**Comment by frouzbeh** [comment]
I'm facing the same issue. No osd is created with raw partition and the osd prepare log shows:
`/dev/sda1 is a partition, please pass LVs or raw block devices
`
rook version: 1.7.3
ceph version: 16.2.5

**Reply by R-omk** [reply]
My related question https://github.com/rook/rook/issues/8494

**Comment by davralin** [comment]
From the original issue:

> 1. Use full devices (e,g: `/dev/sda`) or Logical Volumes on top of your partition, Rook could happily propose LVs to ceph-volume but passing `dm-` does not seem to work on 15.2.9, see: https://tracker.ceph.com/issues/49582
> 2. For the time being, **stick** with v14.2.12 or v15.2.7, and once Rook 1.6 is released, upgrade to **at least** v14.2.14 or v15.2.9 to get partitions working again. As of 1.6, Rook OSD's implementation for simple scenarios (one OSD = one disk basically) is not using LVM but RAW mode from ceph-volume. See https://github.com/rook/rook/pull/4879 for more details.


The options in 2. doesn't work - I'm currently on rook 1.7.8 - with ceph 16.2.6 defined in cluster.yaml - and upon prepare I'm still getting: `ceph-volume lvm batch: error: /dev/sda4 is a partition, please pass LVs or raw block devices`.

Is this supposed to work with these versions?

I also tried, according to 1., create a LV from the partition - but that's ignored during the prepare-steps - even after a boot of that host.

**Reply by davralin** [reply]
No change with 1.8.0.

**Reply by travisn** [reply]
Rook's CI tests creating OSDs on a partition (without a metadata device), so I would expect partitions to be generally working. 

@davralin Are you using a metadataDevice as also mentioned by @mikabytes? If so, it sounds like there is an issue with "batch" mode in ceph-volume where this is failing. 

Another way to use the metadata device is with the [storageClassDeviceSets](https://rook.github.io/docs/rook/v1.8/ceph-cluster-crd.html#dedicated-metadata-and-wal-device-for-osd-on-pvc). This is anyway a better solution since a single metadata device PV would be created for each data PV. If a single metadata device fails it only affects that single OSD. But when the `metadataDevice` is specified with the non-PV approach, the same metadata device will be used for multiple OSDs and be a single point of failure.

**Reply by yydzhou** [reply]
@travisn does that mean the OSD on partition should be working. We have lots of ceph clusters and are hesitating upgrade due to this limitation. However, we have stick to the 15.2.7 for too long and have to catch up. 
Is there any info about in which  rook/ceph version does this get fixed?

**Reply by mikabytes** [reply]
@travisn Thanks for the hint. I will definitely be looking into storageClassDeviceSets :+1: 

Meanwhile, I just now managed to get it working by creating a new PV/VG/LV lvm volume inside the partition. Ceph was happy with that as long as I **did not** specify the full path. 

`/dev/mapper/vg0-lv0` does **not** work
`/dev/vg0/lv0` does **not** work
`vg0/lv0` does work

That being said, this is only for the `metadataDevice`. Earlier today I tried using LVs for data devices with no success, rook didn't find/match the drive. However, I didn't try with the naming convention that I got working now with metadata. Perhaps it would work.

**Reply by travisn** [reply]
> @travisn does that mean the OSD on partition should be working. We have lots of ceph clusters and are hesitating upgrade due to this limitation. However, we have stick to the 15.2.7 for too long and have to catch up. Is there any info about in which rook/ceph version does this get fixed?

@yydzhou Are you using the metadataDevice, or without them? I know the partitions are working on the latest releases without metadata as also verified by the CI. The question is just when the metadataDevice is specified.

**Reply by travisn** [reply]
@mikabytes We have the old issue #2047 to enable using the LVs so I'm not aware of it working in Rook unless there was some change in ceph-volume recently that would allow it to work now.

**Reply by yydzhou** [reply]
@travisn I didn't use metadataDevice. So I guess I am good to go with the new version. I will try it out. Thanks.

**Reply by davralin** [reply]
> @davralin Are you using a metadataDevice as also mentioned by @mikabytes? If so, it sounds like there is an issue with "batch" mode in ceph-volume where this is failing.

The only thing I can find in my config related to that, is the defaults in [cluster.yaml](https://github.com/rook/rook/blob/v1.8.0/deploy/examples/cluster.yaml#L216) - which are commented.
I do however set a deviceClass for each device, in it's config - if that's relevant.

**Reply by travisn** [reply]
@davralin Are you perhaps specifying to encrypt the osds? I believe that will also trigger the "batch" mode where you are seeing it failing. 

To confirm the partitions are working, I just created minikube with a device, created a partition, then specified the device with a deviceFilter in the cluster CR. The OSD was created with "raw" mode without an issue.

**Reply by davralin** [reply]
> @davralin Are you perhaps specifying to encrypt the osds? I believe that will also trigger the "batch" mode where you are seeing it failing.

Oh yes, that I am!
[cluster.yaml](https://github.com/rook/rook/blob/v1.8.0/deploy/examples/cluster.yaml#L220) - `encryptedDevice: "true"` is uncommented.

If needed, I can try setting the value to false and re-running Prepare to confirm?

Can probably do that some time tomorrow, if needed.

**Reply by travisn** [reply]
Ok, that should do it. So the summary is that OSDs can't be created on partitions with either encryption or metadata device unless using storageClassDeviceSets. @leseb Anything else to add?

**Reply by leseb** [reply]
No, I think that's it, if there is something to be changed, it's in ceph-volume anyways.

**Reply by davralin** [reply]
I'm not sure if I could provide any more info, but fwiw - the issue remains with 1.8.1.

**Reply by davralin** [reply]
@travisn not sure if you really needed any more debugging, or if I should separate this into an issue in itself - just wanted to let you know that the issue is present in 1.8.5 also.

A little bit hesitant to create an issue for this, since it seems you already know where the fault lies - but a bit afraid this is forgotten deep in a discussion too...

**Reply by travisn** [reply]
The functionality is there today, but the catch is that you would need to create local PVs to make those partitions available, then use a `storageClassDeviceSet` to create the OSDs on the local PVs. For example, you can create static PVs like [this CI script](https://github.com/rook/rook/blob/master/tests/scripts/localPathPV.sh#L124-L163), or you could create PVs more dynamically with something like the [LSO](https://github.com/openshift/local-storage-operator/blob/master/docs/deploy-with-olm.md). Have you looked into these options? Updating rook and ceph-volume to support these scenarios for non-PVCs just hasn't been a priority.

**Reply by davralin** [reply]
Ahh, right. From this discussion I assumed you considered it an easy fix to just get out of the way - re-reading I see that I misinterpreted that.

I know my use-case is an edge-case, so for the sake of going forward I went ahead and set `encryptedDevice` to `false` - after re-running prepare (and perhaps clearing the partition with wipefs), the osd was instantly created on a partition.

If anything, use this as a confirmation.

**Comment by mikabytes** [comment]
I'm also seeing this, running Rook 1.8.0 with Ceph 16.2.7

My partition is for metadata. Same error as other people mentioned above.

**Comment by grafjo** [comment]
is there a way to run use a partition for an OSD with encryption enabled?
